pkg/storage/closedts/minprop/tracker.go

```diff
@@ -212,50 +224,82 @@ func (t *Tracker) Track(
 	}
 
 	var calls int
-	release := func(ctx context.Context, rangeID roachpb.RangeID, lai ctpb.LAI) {
+	release := func(ctx context.Context, epoch ctpb.Epoch, rangeID roachpb.RangeID, lai ctpb.LAI) {
 		calls++
 		if calls != 1 {
 			if lai != 0 || rangeID != 0 || calls > 2 {
 				log.Fatal(ctx, log.Safe(fmt.Sprintf("command released %d times, this time with arguments (%d, %d)", calls, rangeID, lai)))
 			}
 			return
 		}
+		t.untrack(ctx, minProp, epoch, rangeID, lai, shouldLog)
+	}
 
-		t.mu.Lock()
-		defer t.mu.Unlock()
-		if minProp == t.mu.closed.Next() {
-			if shouldLog {
-				log.Infof(ctx, "release: minprop %s on r%d@%d tracked on the left", minProp, rangeID, lai)
-			}
-			t.mu.leftRef--
-			if t.mu.leftRef < 0 {
-				log.Fatalf(ctx, "min proposal left ref count < 0")
-			}
-			if rangeID == 0 {
-				return
-			}
+	return minProp, release
+}
 
+func clearMLAIMap(m map[roachpb.RangeID]ctpb.LAI) {
+	for rangeID := range m {
+		delete(m, rangeID)
+	}
+}
+
+func (t *Tracker) untrack(
+	ctx context.Context,
+	minProp hlc.Timestamp,
+	epoch ctpb.Epoch,
+	rangeID roachpb.RangeID,
+	lai ctpb.LAI,
+	shouldLog bool,
+) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// If the epoch has increased we need to clear the current vie of the MLAI
+	// maps as they pertain to a previous epoch. Tracking LAI relative to the
+	// latest epoch allows the Tracker to provide a client with information
+	if epoch > t.mu.maxEpoch {
+		t.mu.maxEpoch = epoch
+		for rangeID := range t.mu.leftMLAI {
+			delete(t.mu.leftMLAI, rangeID)
+		}
+		for rangeID := range t.mu.rightMLAI {
+			delete(t.mu.rightMLAI, rangeID)
+		}
+	}
+	fromPrevEpoch := epoch < t.mu.maxEpoch
+	if minProp == t.mu.closed.Next() {
+		if shouldLog {
+			log.Infof(ctx, "release: minprop %s on r%d@%d tracked on the left", minProp, rangeID, lai)
+		}
+		t.mu.leftRef--
+		if t.mu.leftRef < 0 {
+			log.Fatalf(ctx, "min proposal left ref count < 0")
+		}
+		if rangeID == 0 {
+			return
+		}
+		if !fromPrevEpoch {
```

Comment:
I'm curious if I put this into the if condition whether the go compiler would be smart enough to reorder the check and not do the map read if the condition is false

Reply:
Seems like no: https://go.godbolt.org/z/nxU-rC